src/anndata/experimental/backed/_lazy_arrays.py

```diff
@@ -0,0 +1,155 @@
+from __future__ import annotations
+
+from functools import singledispatchmethod
+from typing import TYPE_CHECKING, Generic, TypeVar, Union
+
+import pandas as pd
+
+from anndata._core.index import _subset
+from anndata._core.views import as_view
+from anndata.compat import H5Array, ZarrArray
+
+from ..._settings import settings
+from ._compat import BackendArray, DataArray, ZarrArrayWrapper, xr
+
+if TYPE_CHECKING:
+    from anndata._core.index import Index
+
+
+K = TypeVar("K", bound=Union[H5Array, ZarrArray])
+
+
+class ZarrOrHDF5Wrapper(ZarrArrayWrapper, Generic[K]):
+    @singledispatchmethod  # type: ignore
+    def __init__(self, array: ZarrArray):
+        return super().__init__(array)
+
+    @__init__.register
+    def _(self, array: H5Array):
+        self._array = array
+        self.shape = self._array.shape
+        self.dtype = self._array.dtype
+
+    def __getitem__(self, key):
+        if isinstance(self._array, ZarrArray):
+            return super().__getitem__(key)
+        # adapted from https://github.com/pydata/xarray/blob/main/xarray/backends/h5netcdf_.py#L50-L58C13
+        # TODO: locks?
+        return xr.core.indexing.explicit_indexing_adapter(
+            key,
+            self.shape,
+            xr.core.indexing.IndexingSupport.OUTER_1VECTOR,
+            lambda key: self._array[key],
+        )
+
+
+# Prevents first access from having to load the categories array
+class CategoricalDtypeAccessor:
+    def __init__(self, categories: ZarrArray | H5Array, ordered: bool):
+        self._categories = categories
+        self._ordered = ordered
+        self._dtype = None
+
+    def __get__(self, obj, objtype=None):
+        return self.dtype
+
+    @property
+    def dtype(self):
+        if self._dtype is None:
+            self._dtype = pd.CategoricalDtype(
+                categories=self._categories, ordered=self._ordered
+            )
+        return self._dtype
+
+    def __getattr__(self, name: str):
+        return getattr(self.dtype, name)
+
+    def __repr__(self):
+        return repr(self.dtype)
+
+    def __str__(self) -> str:
+        return str(self.dtype)
+
+
+class CategoricalArray(BackendArray):
+    def __init__(
+        self,
+        codes: ZarrArray | H5Array,
+        categories: ZarrArray | H5Array,
+        ordered: bool,
+        *args,
+        **kwargs,
+    ):
+        self._categories = categories
+        self._ordered = ordered
+        self._categories_cache = None
+        self._codes = ZarrOrHDF5Wrapper[type(codes)](codes)
+        self.shape = self._codes.shape
+        self.dtype = CategoricalDtypeAccessor(
+            categories=self._categories, ordered=self._ordered
+        )
+
+    @property
+    def categories(self):  # __slots__ and cached_property are incompatible
```

Comment:
I don’t see any `__slots__`. Does `BackendArray` define it? And doesn’t subclassing allow to have a `__dict__` and therefore `cached_property`s?

Reply:
Ah, so I think this is again a result of old code, when we had slots, but not 100% sure.  If it breaks, we'll see and can revert.